File: src/services/fee-service.ts
/**
 * Fee calculation.
 *
 *   fee = max(minimum_charge, ceil(minutes / 60) * hourly_rate)
 *
 * with an optional daily cap (defaults to 24h). Hourly rates and the
 * minimum charge are configurable; rates are keyed on `VehicleType`, so
 * adding a new vehicle class is a one-line change.
 */

import { getSettings, type Settings } from "../config.js";
import { VehicleType } from "../types/enums.js";

export interface FeeQuote {
  durationMinutes: number;
  chargeableHours: number;
  ratePerHour: number;
  amount: number;
}

export class FeeCalculator {
  private readonly settings: Settings;

  constructor(settings?: Settings) {
    this.settings = settings ?? getSettings();
  }

  private rateFor(vt: VehicleType): number {
    switch (vt) {
      case VehicleType.MOTORCYCLE:
        return this.settings.feeMotorcyclePerHour;
      case VehicleType.CAR:
        return this.settings.feeCarPerHour;
      case VehicleType.BUS:
        return this.settings.feeBusPerHour;
    }
  }

  /** Parse incoming datetimes (ISO string or Date) into a UTC epoch ms. */
  private toEpochMs(value: string | Date): number {
    const ms =
      typeof value === "string" ? Date.parse(value) : value.getTime();
    if (!Number.isFinite(ms)) {
      throw new Error(`Invalid datetime: ${String(value)}`);
    }
    return ms;
  }

  quote(args: {
    vehicleType: VehicleType;
    entryTime: string | Date;
    exitTime: string | Date;
  }): FeeQuote {
    const entry = this.toEpochMs(args.entryTime);
    const exit = this.toEpochMs(args.exitTime);
    if (exit < entry) {
      throw new Error("exitTime must be >= entryTime");
    }

    const deltaSeconds = (exit - entry) / 1000;
    // Round duration up to the next whole minute — never under-charge by a
    // few seconds.
    const durationMinutes = Math.max(0, Math.ceil(deltaSeconds / 60));

    // Apply minimum chargeable duration.
    const chargeableMinutes = Math.max(
      durationMinutes,
      this.settings.feeMinimumMinutes,
    );

    // Round hours up — a 1h05m stay pays for 2h.
    let chargeableHours = Math.max(1, Math.ceil(chargeableMinutes / 60));
    chargeableHours = Math.min(chargeableHours, this.settings.feeDailyCapHours);

    const rate = this.rateFor(args.vehicleType);
    const amount = Math.round(chargeableHours * rate * 100) / 100;

    return { durationMinutes, chargeableHours, ratePerHour: rate, amount };
  }
}

File: app/services/fee_service.py
"""Fee calculation.

The strategy is intentionally simple and easy to swap out:

  fee = max(minimum_charge, ceil(minutes / 60) * hourly_rate)

with an optional daily cap (by default, 24h). Hourly rates and the minimum
charge are configurable; rates are keyed on :class:`VehicleType` so adding a
new vehicle class is a one-line change.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timezone

from app.config import Settings, get_settings
from app.models.enums import VehicleType


@dataclass(frozen=True)
class FeeQuote:
    duration_minutes: int
    chargeable_hours: int
    rate_per_hour: float
    amount: float


class FeeCalculator:
    """Compute parking fees from entry/exit times and vehicle type."""

    def __init__(self, settings: Settings | None = None) -> None:
        self._settings = settings or get_settings()

    def _rate_for(self, vt: VehicleType) -> float:
        s = self._settings
        return {
            VehicleType.MOTORCYCLE: s.fee_motorcycle_per_hour,
            VehicleType.CAR: s.fee_car_per_hour,
            VehicleType.BUS: s.fee_bus_per_hour,
        }[vt]

    @staticmethod
    def _to_utc(dt: datetime) -> datetime:
        """Normalize naive datetimes to UTC (assume server clock is UTC)."""
        if dt.tzinfo is None:
            return dt.replace(tzinfo=timezone.utc)
        return dt.astimezone(timezone.utc)

    def quote(
        self,
        *,
        vehicle_type: VehicleType,
        entry_time: datetime,
        exit_time: datetime,
    ) -> FeeQuote:
        # Normalize to UTC first so we can compare safely — SQLite ignores
        # timezones and returns naive datetimes, while in-memory values are
        # timezone-aware.
        entry = self._to_utc(entry_time)
        exit_ = self._to_utc(exit_time)

        if exit_ < entry:
            raise ValueError("exit_time must be >= entry_time")

        delta_seconds = (exit_ - entry).total_seconds()
        # Round duration up to the next whole minute so we never under-charge
        # by a few seconds.
        duration_minutes = max(0, math.ceil(delta_seconds / 60))

        # Apply minimum chargeable duration (e.g. a 2-minute stay still pays).
        chargeable_minutes = max(duration_minutes, self._settings.fee_minimum_minutes)

        # Round hours UP — a 1h05m stay is billed as 2 hours.
        chargeable_hours = max(1, math.ceil(chargeable_minutes / 60))

        # Apply daily cap.
        chargeable_hours = min(chargeable_hours, self._settings.fee_daily_cap_hours)

        rate = self._rate_for(vehicle_type)
        amount = round(chargeable_hours * rate, 2)

        return FeeQuote(
            duration_minutes=duration_minutes,
            chargeable_hours=chargeable_hours,
            rate_per_hour=rate,
            amount=amount,
        )
